src/eureka/S1_detector_processing/bias_sub.py

```diff
@@ -0,0 +1,187 @@
+    input_model: data model object
+        science data to be corrected
+
+    bias_model: super-bias model object
+        bias data
+
+    Returns
+    -------
+    output_model: data model object
+        bias-subtracted science data
+
+    """
+    # Check for subarray mode and extract subarray from the
+    # bias reference data if necessary
+    if not reffile_utils.ref_matches_sci(input_model, bias_model):
+        bias_model = reffile_utils.get_subarray_model(input_model, bias_model)
+
+    # Compute trace mask, where ones are good background regions
+    input_model = mask_trace(input_model, log, meta)
+    trace_mask = input_model.trace_mask
+
+    # Compute median of bias frame outside of trace
+    bias_median = np.nanmedian(bias_model.data[np.where(trace_mask)])
+
+    log.writelog('  Computing scale factor for superbias correction.')
+    # Compute median of each integration/group outside of trace
+    # Using double-for loop to avoid heavy memory usage
+    nint, ngroup, nrow, ncol = input_model.data.shape
+    scale_factor = np.zeros((nint, ngroup))
+    for ii in range(nint):
+        data = input_model.data[ii]
+        for jj in range(ngroup):
+            data_median = np.nanmedian(data[jj][np.where(trace_mask)])
+            scale_factor[ii, jj] = data_median/bias_median
+    mean_scale_factor = np.mean(scale_factor, axis=0)
+    log.writelog(f'  Mean bias scale factors (by group): {mean_scale_factor}')
+    
+    # Get segment number
+    segment = str(input_model.meta.exposure.segment_number).zfill(3)
+
+    # Save scale factor to ECSV file
+    fname = meta.outputdir+"S1_seg"+segment+"_BiasScaleFactor.ecsv"
+    savetable_S1(fname, scale_factor)
+
+    # Smooth scale factor
+    if meta.bias_correction == "smooth":
+        for jj in range(ngroup):
+            scale_factor[:, jj] = smooth(scale_factor[:, jj], 
+                                         meta.bias_smooth_length)
+        # Save smoothed scale factor to ECSV file
+        fname = meta.outputdir+"S1_seg"+segment+"_BiasSmoothedScaleFactor.ecsv"
+        savetable_S1(fname, scale_factor)
+
+    # Replace NaN's in the superbias with zeros
+    bias_model.data[np.isnan(bias_model.data)] = 0.0
+
+    # Create 4D bias array for each integration/group and apply scale factor 
+    bias_data = np.ones((nint, ngroup, nrow, ncol)) * \
+        bias_model.data[np.newaxis, np.newaxis, :, :]
+    
+    try:
+        if meta.bias_correction == "mean":
+            if isinstance(meta.bias_group, int):
+                # Scale superbias frame using single value for all groups
+                jj = meta.bias_group
+                assert jj > 0, f"Group number should be >0, got: {jj}"
+                assert jj <= ngroup, f"Group number should be <={ngroup}" + \
+                    ", got: {jj}"
+                log.writelog('  Applying mean bias correction using ' +
+                             f'group {jj}.')
+                bias_model.data *= mean_scale_factor[jj-1]
```

Comment:
Here you only modufy `bias_model.data`, while below you sometimes modify only `bias_data` or modify both of them. I just want to clarify that's your intended behaviour since I don't really understand the differences between the two

Reply:
Nice catch!  I forgot to include `bias_data = None` in this section.  

`bias_model.data` is the calibrated superbias file from commissioning.  It's dimensions (ny, nx) cannot be changed to allow for group- or integration-level variations.

`biad_data` is the 4D equivalent (nint, ngroup, ny, nx) that enables time-dependent corrections.